### PR TITLE
Update irmin to the last version of ocaml-git

### PR DESCRIPTION
**Pull request by dinosaure**
This is a draft about an update to the new version of git presented on mirage/ocaml-git#395 and the new version of conduit presented on mirage/ocaml-conduit#311 (including patches into mirage/ocaml-cohttp#692). I think the patch is small enough. The CI will fail but locally (with a `pin` of `conduit` and `cohttp` and a simple `dune runtest`) it's work (believe me).

This is a draft but people can start to comment it if some details are obscures. Currently, the biggest change is about the synchronization where I updated the interface which seems to be much more simpler.

### CODE REVIEW

**Comment by CraigFe** [comment]
> The CI will fail but locally (with a `pin` of `conduit` and `cohttp` and a simple `dune runtest`) it's work (believe me).

If you want to get the CI to work, you can add `pin-depends` on specific commits as we're doing [here](https://github.com/mirage/irmin/blob/master/irmin-pack.opam#L27-L29). (Needs to be a commit hash, otherwise OCaml-CI will not accept it.) Also fine to just leave this failing until the upstream changes have been released :shrug:

**Comment by dinosaure** [comment]
- [ ] clean the PR
- [x] CI is happy

**Comment by dinosaure** [comment]
Re-updated to the last version of `ocaml-git` and remove the update to `conduit.3.0.0` (useless due to mirage/ocaml-git#428)

**Comment by dinosaure** [comment]
Ready to review :+1:., `ocaml-git` is freezed.

File: irmin-graphql.opam [hunk 1]
```diff
@@ -25,6 +25,7 @@ depends: [
   "cohttp"
   "fmt"
   "lwt"
+  "mimic"
```
Comment (samoht) [inline irmin-graphql.opam:28]: why this is leaking here? `irmin-grapql` has nothing to do with ocaml-git.

File: src/irmin-graphql/dune [hunk 2]
```diff
@@ -1,5 +1,5 @@
 (library
  (name irmin_graphql)
  (public_name irmin-graphql)
- (libraries fmt cohttp cohttp-lwt graphql-cohttp graphql graphql-lwt
+ (libraries mimic fmt cohttp cohttp-lwt graphql-cohttp graphql graphql-lwt
```
Comment (samoht) [inline src/irmin-graphql/dune:4]: same question here ; I'm not super happy to see this leaking in unrelated code.

**Comment by samoht** [comment]
Thanks for the patch. My main concern is that I would like to avoid `mimic` to leak in non-git related codebase. It also seems to me that `Mimic` should be renamed `Git_conduit` to avoid confusion (as it's a conduit clone only used by ocaml-git).

File: src/irmin-graphql/dune (hunk 2 above)
Comment (dinosaure) [reply src/irmin-graphql/dune:4]: Seems like an bad `sed` over `conduit` and `mimic`.

File: src/irmin-mirage/graphql/irmin_mirage_graphql.ml [hunk 3]
```diff
@@ -4,14 +4,17 @@ module Server = struct
 
     module Http : Cohttp_lwt.S.Server
 
-    module Store : Irmin.S with type Private.Sync.endpoint = Git_mirage.endpoint
```
Comment (dinosaure) [inline src/irmin-mirage/graphql/irmin_mirage_graphql.ml:7]: I think the insertion of `Mimic` on graphql comes here. I don't understand why `irmin-mirage-graphql` decided to use `Git_mirage.endpoint`. On my side, it sure that a `Git_mirage.endpoint/Smart.Endpoint.t` must be associated to a `Mimic.ctx`. I'm not use how to solve this issue however.

**Comment by dinosaure** [comment]
> It also seems to me that Mimic should be renamed Git_conduit to avoid confusion (as it's a conduit clone only used by ocaml-git).

I'm not sure, even if `mimic` seems specialized to `ocaml-git`, I move some of my codebase (SMTP and HTTP/AF) with this small library.

File: irmin-git.opam [hunk 4]
```diff
@@ -18,7 +18,7 @@ depends: [
   "dune"       {>= "2.7.0"}
   "irmin"      {= version}
   "ppx_irmin"  {= version}
-  "git"        {>= "2.1.1"}
+  "git"
```
Comment (samoht) [inline irmin-git.opam:21]: any reason to remove the constraint here?

Comment (dinosaure) [reply irmin-git.opam:21]: Will be `>= 3.0.0` at the release of `ocaml-git`.

File: irmin-http.opam [hunk 5]
```diff
@@ -28,11 +28,32 @@ depends: [
...
   "irmin-mem"  {with-test & = version}
   "irmin-test" {with-test & = version}
   "git-unix"   {with-test}
   "digestif"   {with-test & >= "0.9.0"}
+  "alcotest"   {>= "1.2.3" & with-test}
+  "cohttp-lwt-unix" {>= "2.5.4" & with-test}
+  "conduit-lwt-unix" {>= "2.2.2" & with-test}
+  "fpath"      {>= "0.7.3" & with-test}
+  "git"        {>= "2.1.3" & with-test}
+  "git-cohttp-unix" {>= "~dev" & with-test}
```
Comment (samoht) [inline irmin-http.opam:42]: why all these extra test dependencies?

File: src/irmin-git/irmin_git.ml [hunk 6]
```diff
@@ -747,70 +736,76 @@ struct
 
   type branch = B.t
 
-  type endpoint = S.Endpoint.t
+  type endpoint = Mimic.ctx * Smart_git.Endpoint.t
```
Comment (samoht) [inline src/irmin-git/irmin_git.ml:739]: Would it be possible to hide `Mimic.ctx` into `Smart_git.Endpoint.t`? Why does it need to be separated here?

File: src/irmin-git/irmin_git.ml [hunk 7]
```diff
@@ -1055,74 +1037,48 @@ module Mem = struct
...
+  type hash = G.hash
 
-  module Endpoint = struct
-    type t = unit
+  type store = G.t
 
-    let uri _ = assert false
-  end
+  (* XXX(samoht): so much boilerplate... *)
+  (* XXX(dinosaure): not anymore. *)
```
Comment (samoht) [inline src/irmin-git/irmin_git.ml:1063]: maybe just remove these 2 lines :-)

File: irmin-http.opam (hunk 5 above)
Comment (dinosaure) [reply irmin-http.opam:42]: It's a pass of `opam-dune-lint` locally, I don't really why it added these dependencies.

File: src/irmin-unix/graphql.ml [hunk 8]
```diff
@@ -20,7 +20,10 @@ module Server = struct
       (struct
         let info = Info.v
 
-        let remote = Remote.remote
+        let remote =
+          match Remote.remote with
+          | Some fn -> Some (fun ?headers v -> fn ?headers v)
```
Comment (samoht) [inline src/irmin-unix/graphql.ml:25]: is this still needed (the pattern-match looks like a no-op)?

File: src/irmin-git/irmin_git.ml (hunk 6 above)
Comment (dinosaure) [reply src/irmin-git/irmin_git.ml:739]: It's not really the same purpose and `Mimic.ctx` can be used with something else than a `Smart_git.Endpoint.t` (e.g. an `Uri.t` for HTTP/AF for example).

File: irmin-http.opam (hunk 5 above)
Comment (samoht) [reply irmin-http.opam:42]: it's weird, `opam-dune-lint` doesn't pick these on master.

File: src/irmin-unix/graphql.ml (hunk 8 above)
Comment (dinosaure) [reply src/irmin-unix/graphql.ml:25]: It's seems that `Remote.remote` comes from `irmin-git` which expects an `?ctx` argument. The no-op erases the `?ctx` argument which is not used (according to https://github.com/mirage/irmin/pull/1065#discussion_r553243932) by `irmin-graphql`.

File: src/irmin-git/irmin_git.ml (hunk 6 above)
Comment (samoht) [reply src/irmin-git/irmin_git.ml:739]: ok makes sense.

**Comment by samoht** [comment]
The diff in `irmin-http` are weird but I can live with that :-)

Could you also add a line in the changelog (and point to https://github.com/mirage/irmin/issues?q=is%3Aissue+is%3Aopen+git+label%3Agit which would probably be fixed :tada:) ? After that I think we are good to merge.

File: irmin-http.opam (hunk 5 above)
Comment (samoht) [reply irmin-http.opam:42]: hum actually I am not sure why we need to depend on `git` and `git-cohttp-unix` here.

Comment (dinosaure) [reply irmin-http.opam:42]: me neither, I can revert this diff if you want, may be I use an old version of `opam-dune-lint`. Let me check.

**Comment by dinosaure** [comment]
> The diff in irmin-http.opam are weird but I can live with that :-) why does irmin-http now depends on Git? I am not sure I am good with this change.

I reverted the patch on `irmin-http.opam` however, `dune-opam-lint` still complains about missing dependencies even if I did not updated `src/irmin-http/dune` or the irmin-unix's codebase (according to the diff). So I don't really understand why such dependencies appears.

**Comment by dinosaure** [comment]
So my revert seems good. However, I don't think that this PR fixes #914 or #917 (but #1001 is fixed according to [a test in `ocaml-git`](https://github.com/mirage/ocaml-git/blob/25043a26b935a77af476617ea07b0408841ab72d/test/tree/test.ml#L92-L99)). I will add an entry into the `Changelog` :+1:.

**Comment by dinosaure** [comment]
According to the release of `git.3.0.0` (see ocaml/opam-repository#17950), this PR is ready to merge.

**Comment by samoht** [comment]
Can you fix `dune build @fmt` to make the CI happy?

File: irmin-unix.opam [hunk 9]
```diff
@@ -40,15 +41,9 @@ depends: [
...
   "fmt"
-  "git"
+  "git"             {>= "3.0.0"}
+  "git-cohttp-unix" {>= "3.0.0"}
   "lwt"
-  "irmin-test" {with-test & = version}
-  "alcotest"   {with-test}
+  "irmin-test"    {with-test & = version}
+  "alcotest"      {with-test}
 ]
```
Comment (samoht) [inline irmin-unix.opam:49]: Can you also re-add the missing description and synopsis? Thanks!

**Comment by dinosaure** [comment]
It's done :+1:.

**Comment by CraigFe** [comment]
LGTM, thanks!
